File: tests_ext/src/test/java/com/dirgub/klee0kai/stone/text_ext/qualifiers/GoodPhoneInjectExtTests.java
package com.dirgub.klee0kai.stone.text_ext.qualifiers;

import com.github.klee0kai.stone.Stone;
import com.github.klee0kai.test.di.base_phone.PhoneComponent;
import com.github.klee0kai.test.di.base_phone.qualifiers.DataStorageSize;
import com.github.klee0kai.test.di.base_phone.qualifiers.RamSize;
import com.github.klee0kai.test.tech.phone.GoodPhone;
import com.github.klee0kai.test_ext.inject.di.base_phone.PhoneExtComponent;
import com.github.klee0kai.test_ext.inject.tech.components.DDR3Ram;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class GoodPhoneInjectExtTests {


    @Test
    @Disabled("https://github.com/klee0kai/stone/issues/42")
    public void createdIsReusableTest() {
        //Given
        PhoneComponent DI = Stone.createComponent(PhoneComponent.class);
        GoodPhone goodPhone1 = new GoodPhone();
        DI.inject(goodPhone1, goodPhone1.lifeCycleOwner, new DataStorageSize("120GB"), new RamSize("8GB"));

        //When
        PhoneExtComponent DIPro = Stone.createComponent(PhoneExtComponent.class);
        DIPro.extOf(DI);
        GoodPhone goodPhone2 = new GoodPhone();
        DIPro.inject(goodPhone2, goodPhone2.lifeCycleOwner, new DataStorageSize("120GB"), new RamSize("8GB"));

        //Then
        assertEquals(goodPhone1.ram.uuid, goodPhone2.ram.uuid);
        assertTrue(goodPhone1.ram instanceof DDR3Ram);
    }


    @Test
    public void createGoodPhoneFromProtoTest() {
        //Given
        PhoneComponent DI = Stone.createComponent(PhoneComponent.class);
        PhoneExtComponent DIPro = Stone.createComponent(PhoneExtComponent.class);
        DIPro.extOf(DI);

        GoodPhone goodPhone = new GoodPhone();

        //When
        DI.inject(goodPhone, goodPhone.lifeCycleOwner, new DataStorageSize("120GB"), new RamSize("8GB"));

        //Then
        assertEquals("120GB", goodPhone.dataStorage.size);
        assertEquals("8GB", goodPhone.ram.size);
        assertTrue(goodPhone.ram instanceof DDR3Ram);
    }


    @Test
    public void createGoodPhoneFromProTest() {
        //Given
        PhoneComponent DI = Stone.createComponent(PhoneComponent.class);
        PhoneExtComponent DIPro = Stone.createComponent(PhoneExtComponent.class);
        DIPro.extOf(DI);

        GoodPhone goodPhone = new GoodPhone();

        //When
        DIPro.inject(goodPhone, goodPhone.lifeCycleOwner, new DataStorageSize("120GB"), new RamSize("8GB"));

        //Then
        assertEquals("120GB", goodPhone.dataStorage.size);
        assertEquals("8GB", goodPhone.ram.size);
        assertTrue(goodPhone.ram instanceof DDR3Ram);
    }


    @Test
    public void createGoodPhoneFromProExtTest() {
        //Given
        PhoneComponent DI = Stone.createComponent(PhoneComponent.class);
        PhoneExtComponent DIPro = Stone.createComponent(PhoneExtComponent.class);
        DIPro.extOf(DI);
        GoodPhone goodPhone = new GoodPhone();

        //When
        DIPro.injectExt(goodPhone, goodPhone.lifeCycleOwner, new DataStorageSize("120GB"), new RamSize("8GB"));

        //Then
        assertEquals("120GB", goodPhone.dataStorage.size);
        assertEquals("8GB", goodPhone.ram.size);
        assertTrue(goodPhone.ram instanceof DDR3Ram);
    }


    @Test
    public void cacheInjectedPhoneTest() {
        //Given
        PhoneComponent DI = Stone.createComponent(PhoneComponent.class);
        PhoneExtComponent DIPro = Stone.createComponent(PhoneExtComponent.class);
        DIPro.extOf(DI);
        GoodPhone goodPhone1 = new GoodPhone();
        GoodPhone goodPhone2 = new GoodPhone();


        //When
        DI.inject(goodPhone1, goodPhone1.lifeCycleOwner, new DataStorageSize("120GB"), new RamSize("8GB"));
        DIPro.inject(goodPhone2, goodPhone2.lifeCycleOwner, new DataStorageSize("120GB"), new RamSize("8GB"));

        //Then
        assertEquals(goodPhone1.ram.uuid, goodPhone2.ram.uuid);
        assertTrue(goodPhone1.ram instanceof DDR3Ram);
    }

}

File: tests_ext/src/test/java/com/dirgub/klee0kai/stone/text_ext/identifiers/GoodPhoneInjectExtTests.kt
package com.dirgub.klee0kai.stone.text_ext.identifiers

import com.github.klee0kai.test.di.base_phone.PhoneComponent
import com.github.klee0kai.test.di.base_phone.PhoneComponentStoneComponent
import com.github.klee0kai.test.di.base_phone.identifiers.DataStorageSize
import com.github.klee0kai.test.di.base_phone.identifiers.RamSize
import com.github.klee0kai.test.tech.phone.GoodPhone
import com.github.klee0kai.test_ext.inject.di.base_phone.PhoneExtComponent
import com.github.klee0kai.test_ext.inject.di.base_phone.PhoneExtComponentStoneComponent
import com.github.klee0kai.test_ext.inject.tech.components.DDR3Ram
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Test

class GoodPhoneInjectExtTests {


    @Test
    @Disabled("https://github.com/klee0kai/stone/issues/42")
    fun createdIsReusableTest() {
        //Given
        val DI: PhoneComponent = PhoneComponentStoneComponent()
        val goodPhone1 = GoodPhone()
        DI.inject(goodPhone1, goodPhone1.lifeCycleOwner, DataStorageSize("120GB"), RamSize("8GB"))

        //When
        val DIPro: PhoneExtComponent = PhoneExtComponentStoneComponent()
        DIPro.extOf(DI)
        val goodPhone2 = GoodPhone()
        DIPro.inject(goodPhone2, goodPhone2.lifeCycleOwner, DataStorageSize("120GB"), RamSize("8GB"))

        //Then
        assertEquals(goodPhone1.ram!!.uuid, goodPhone2.ram!!.uuid)
        assertTrue(goodPhone1.ram is DDR3Ram)
    }


    @Test
    fun createGoodPhoneFromProtoTest() {
        //Given
        val DI: PhoneComponent = PhoneComponentStoneComponent()
        val DIPro: PhoneExtComponent = PhoneExtComponentStoneComponent()
        DIPro.extOf(DI)

        val goodPhone = GoodPhone()

        //When
        DI.inject(goodPhone, goodPhone.lifeCycleOwner, DataStorageSize("120GB"), RamSize("8GB"))

        //Then
        assertEquals("120GB", goodPhone.dataStorage!!.size)
        assertEquals("8GB", goodPhone.ram!!.size)
        assertTrue(goodPhone.ram is DDR3Ram)
    }


    @Test
    fun createGoodPhoneFromProTest() {
        //Given
        val DI: PhoneComponent = PhoneComponentStoneComponent()
        val DIPro: PhoneExtComponent = PhoneExtComponentStoneComponent()
        DIPro.extOf(DI)

        val goodPhone = GoodPhone()

        //When
        DIPro.inject(goodPhone, goodPhone.lifeCycleOwner, DataStorageSize("120GB"), RamSize("8GB"))

        //Then
        assertEquals("120GB", goodPhone.dataStorage!!.size)
        assertEquals("8GB", goodPhone.ram!!.size)
        assertTrue(goodPhone.ram is DDR3Ram)
    }


    @Test
    fun createGoodPhoneFromProExtTest() {
        //Given
        val DI: PhoneComponent = PhoneComponentStoneComponent()
        val DIPro: PhoneExtComponent = PhoneExtComponentStoneComponent()
        DIPro.extOf(DI)
        val goodPhone = GoodPhone()

        //When
        DIPro.injectExt(goodPhone, goodPhone.lifeCycleOwner, DataStorageSize("120GB"), RamSize("8GB"))

        //Then
        assertEquals("120GB", goodPhone.dataStorage!!.size)
        assertEquals("8GB", goodPhone.ram!!.size)
        assertTrue(goodPhone.ram is DDR3Ram)
    }


    @Test
    fun cacheInjectedPhoneTest() {
        //Given
        val DI: PhoneComponent = PhoneComponentStoneComponent()
        val DIPro: PhoneExtComponent = PhoneExtComponentStoneComponent()
        DIPro.extOf(DI)
        val goodPhone1 = GoodPhone()
        val goodPhone2 = GoodPhone()


        //When
        DI.inject(goodPhone1, goodPhone1.lifeCycleOwner, DataStorageSize("120GB"), RamSize("8GB"))
        DIPro.inject(goodPhone2, goodPhone2.lifeCycleOwner, DataStorageSize("120GB"), RamSize("8GB"))

        //Then
        assertEquals(goodPhone1.ram!!.uuid, goodPhone2.ram!!.uuid)
        assertTrue(goodPhone1.ram is DDR3Ram)
    }

}
